include/hve/arch/intel_x64/irq_window.h

```diff
@@ -0,0 +1,141 @@
+//
+// Bareflank Extended APIs
+//
+// Copyright (C) 2018 Assured Information Security, Inc.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#ifndef IRQWINDOW_HDLR_INTEL_X64_EAPIS_H
+#define IRQWINDOW_HDLR_INTEL_X64_EAPIS_H
+
+#include <list>
+
+#include <bfgsl.h>
+#include <bfvmm/hve/arch/intel_x64/vmcs/vmcs.h>
+#include <bfvmm/hve/arch/intel_x64/exit_handler/exit_handler.h>
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_EAPIS_HVE
+#ifdef SHARED_EAPIS_HVE
+#define EXPORT_EAPIS_HVE EXPORT_SYM
+#else
+#define EXPORT_EAPIS_HVE IMPORT_SYM
+#endif
+#else
+#define EXPORT_EAPIS_HVE
+#endif
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+///
+/// Interrupt-window exit
+///
+class EXPORT_EAPIS_HVE irq_window
+{
+public:
+
+    using exit_handler_t = bfvmm::intel_x64::exit_handler;
+    using vmcs_t = bfvmm::intel_x64::vmcs;
+    using handler_t = delegate<bool(gsl::not_null<vmcs_t *>)>;
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    irq_window(gsl::not_null<exit_handler_t *> exit_handler);
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    ~irq_window() = default;
+
+    /// Add handler
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param d the handler to call when an interrupt-window exit
+    ///        occurs
+    ///
+    void add_handler(handler_t &&d);
+
+    /// Trap
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param enable enable if true, disable (i.e. passthrough) if false
+    ///
+    void trap(bool enable = true);
+
+    /// Pass-through
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Disable interrupt-window exiting.
+    ///
+    void pass_through();
```

Comment:
not needed. We only support turning the features on.

Reply:
We need to be able to turn this on and off